File: app/src/main/java/com/settlex/android/ui/auth/activity/LoginActivity.java
package com.settlex.android.ui.auth.activity;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.os.Bundle;
import android.text.Editable;
import android.text.Html;
import android.text.InputType;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.MotionEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.settlex.android.R;
import com.settlex.android.data.remote.profile.ProfileService;
import com.settlex.android.databinding.ActivityLoginBinding;
import com.settlex.android.ui.auth.viewmodel.AuthViewModel;
import com.settlex.android.ui.common.util.ProgressLoaderController;
import com.settlex.android.ui.dashboard.DashboardActivity;
import com.settlex.android.ui.info.help.AuthHelpActivity;
import com.settlex.android.utils.network.NetworkMonitor;
import com.settlex.android.utils.string.StringUtil;
import com.settlex.android.utils.ui.StatusBarUtil;
import com.settlex.android.utils.ui.UiUtil;

import java.util.Objects;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class LoginActivity extends AppCompatActivity {
    private boolean isPasswordVisible = false;

    // dependencies
    private ProgressLoaderController progressLoader;
    private ActivityLoginBinding binding;
    private AuthViewModel authViewModel;
    private boolean isConnected = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityLoginBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        authViewModel = new ViewModelProvider(this).get(AuthViewModel.class);
        progressLoader = new ProgressLoaderController(this);

        setupUiActions();
        observeUserState();
        observeNetworkStatus();
        observeLoginAndHandleResult();
    }

    // OBSERVERS =========
    private void observeNetworkStatus() {
        NetworkMonitor.getNetworkStatus().observe(this, isConnected -> {
            if (!isConnected) showNoInternetDialog();
            this.isConnected = isConnected;
        });
    }

    private void observeUserState() {
        authViewModel.getUserAuthStateLiveData().observe(this, currentUser -> {
            if (currentUser != null) {
                // user is logged in
                showLoggedInLayout(currentUser.getPhotoUrl(), currentUser.getDisplayName(), currentUser.getEmail());
                return;
            }
            showLoggedOutLayout();
        });
    }

    private void observeLoginAndHandleResult() {
        authViewModel.getLoginResult().observe(this, result -> {
            if (result != null) {
                switch (result.getStatus()) {
                    case LOADING -> onLoginLoading();
                    case SUCCESS -> onLoginSuccess();
                    case ERROR -> onLoginFailure(result.getMessage());
                }
            }
        });
    }

    private void onLoginLoading() {
        progressLoader.show();
    }

    private void onLoginSuccess() {
        startActivity(new Intent(this, DashboardActivity.class));
        finishAffinity();
        progressLoader.hide();
    }

    private void onLoginFailure(String error) {
        binding.txtError.setText(error);
        binding.txtError.setVisibility(View.VISIBLE);

        progressLoader.hide();
    }


    private void attemptLogin() {
        if (!isConnected) {
            showNoInternetDialog();
            return;
        }

        String email = Objects.requireNonNull(binding.editTxtEmail.getText()).toString().trim();
        String password = Objects.requireNonNull(binding.editTxtPassword.getText()).toString().trim();

        attemptUserLoginWithEmailAndPassword(email, password);
    }

    private void attemptUserLoginWithEmailAndPassword(String email, String password) {
        authViewModel.loginWithEmail(email, password);
    }

    private void showLoggedInLayout(String photoUrl, String displayName, String email) {
        String userDisplayName = "Hi, " + displayName.toUpperCase();
        String userEmail = "(" + email + ")";

        ProfileService.loadProfilePic(photoUrl, binding.userProfile);
        binding.userDisplayName.setText(userDisplayName);
        binding.userEmail.setText(StringUtil.maskEmail(userEmail));
        binding.editTxtEmail.setText(email);
        // Show
        binding.showCurrentUserLayout.setVisibility(View.VISIBLE);
        binding.btnFingerprint.setVisibility(View.VISIBLE);
        binding.btnSwitchAccount.setVisibility(View.VISIBLE);
        // Hide
        binding.logo.setVisibility(View.GONE);
        binding.txtInputLayoutEmail.setVisibility(View.GONE);
        binding.btnSignUp.setVisibility(View.GONE);
    }

    private void showLoggedOutLayout() {
        // sign out
        authViewModel.signOut();

        // Hide
        binding.showCurrentUserLayout.setVisibility(View.GONE);
        binding.btnFingerprint.setVisibility(View.GONE);
        binding.btnSwitchAccount.setVisibility(View.GONE);
        // Show
        binding.txtInputLayoutEmail.setVisibility(View.VISIBLE);
        binding.btnSignUp.setVisibility(View.VISIBLE);
        binding.logo.setVisibility(View.VISIBLE);
    }

    private void showNoInternetDialog() {
        String title = "Network Unavailable";
        String message = "Please check your Wi-Fi or cellular data and try again";

        UiUtil.showSimpleAlertDialog(
                this,
                title,
                message);
    }

    // UI ACTIONS ==========
    private void setupUiActions() {
        StatusBarUtil.setStatusBarColor(this, R.color.white);
        setupEditTextFocusHandler();
        setupAuthActionTexts();
        setupInputValidation();
        setupPasswordVisibilityToggle();
        clearFocusOnLastEditTextField();

        // Click listeners
        binding.btnBackBefore.setOnClickListener(v -> getOnBackPressedDispatcher().onBackPressed());
        binding.btnSignUp.setOnClickListener(v -> navigateTo(SignUpActivity.class));
        binding.btnHelp.setOnClickListener(v -> navigateTo(AuthHelpActivity.class));
        binding.btnSwitchAccount.setOnClickListener(view -> showLoggedOutLayout());
        binding.btnForgotPassword.setOnClickListener(v -> navigateTo(PasswordResetActivity.class));
        binding.btnSignIn.setOnClickListener(v -> attemptLogin());
    }

    private void setupAuthActionTexts() {
        String signUpActionText = "Don't have an account yet? <font color='#0044CC'><br>Click here to register</font>";
        binding.btnSignUp.setText(Html.fromHtml(signUpActionText, Html.FROM_HTML_MODE_LEGACY));

        String switchAccountActionText = "Not you? <br><font color='#0044CC'>Switch Account</font>";
        binding.btnSwitchAccount.setText(Html.fromHtml(switchAccountActionText, Html.FROM_HTML_MODE_LEGACY));
    }

    private void navigateTo(Class<? extends Activity> activityClass) {
        startActivity(new Intent(this, activityClass));
    }

    private void setupEditTextFocusHandler() {
        int focusBgRes =  R.drawable.bg_edit_txt_custom_white_focused;
        int defaultBgRes = R.drawable.bg_edit_txt_custom_gray_not_focused;

        binding.editTxtPassword.setOnFocusChangeListener((v, hasFocus) -> binding.editTxtPasswordBg.setBackgroundResource(hasFocus ? focusBgRes : defaultBgRes));
    }

    private void updateSignInButtonState() {
        String email = Objects.requireNonNull(binding.editTxtEmail.getText()).toString().trim();
        String password = Objects.requireNonNull(binding.editTxtPassword.getText()).toString().trim();

        binding.btnSignIn.setEnabled(!email.isEmpty() && !password.isEmpty());
    }

    private void setupInputValidation() {
        TextWatcher validationWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void afterTextChanged(Editable s) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                binding.txtError.setVisibility(View.GONE);
                updateSignInButtonState();
            }
        };

        binding.editTxtEmail.addTextChangedListener(validationWatcher);
        binding.editTxtPassword.addTextChangedListener(validationWatcher);
        setupPasswordToggleVisibility();
    }

    private void setupPasswordToggleVisibility() {
        binding.editTxtPassword.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void afterTextChanged(Editable s) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                binding.passwordToggle.setVisibility(!TextUtils.isEmpty(s) ? View.VISIBLE : View.GONE);
            }
        });
    }

    private void setupPasswordVisibilityToggle() {
        binding.passwordToggle.setOnClickListener(v -> {
            Typeface currentTypeface = binding.editTxtPassword.getTypeface();
            isPasswordVisible = !isPasswordVisible;

            int inputType = isPasswordVisible ? InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD : InputType.TYPE_TEXT_VARIATION_PASSWORD;

            binding.editTxtPassword.setInputType(InputType.TYPE_CLASS_TEXT | inputType);
            binding.passwordToggle.setImageResource(isPasswordVisible ? R.drawable.ic_visibility_on_filled : R.drawable.ic_visibility_off_filled);

            binding.editTxtPassword.setTypeface(currentTypeface);
            binding.editTxtPassword.setSelection(binding.editTxtPassword.getText().length());
        });
    }

    private void clearFocusOnLastEditTextField() {
        binding.editTxtPassword.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                // Hide the keyboard
                InputMethodManager imm = (InputMethodManager) v.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
                imm.hideSoftInputFromWindow(v.getWindowToken(), 0);

                // Clear focus
                v.clearFocus();
                return true;
            }
            return false;
        });
    }


    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            View v = getCurrentFocus();
            if (v instanceof EditText) {
                Rect outRect = new Rect();
                v.getGlobalVisibleRect(outRect);
                if (!outRect.contains((int) event.getRawX(), (int) event.getRawY())) {
                    v.clearFocus();
                    hideKeyboard(v);
                }
            }
        }
        return super.dispatchTouchEvent(event);
    }

    private void hideKeyboard(View view) {
        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }
}

File: app/src/main/java/com/settlex/android/presentation/auth/login/LoginActivity.kt
package com.settlex.android.presentation.auth.login

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Patterns
import android.view.MotionEvent
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.settlex.android.R
import com.settlex.android.data.exception.AppException
import com.settlex.android.data.remote.profile.ProfileService.loadProfilePhoto
import com.settlex.android.databinding.ActivityLoginBinding
import com.settlex.android.presentation.auth.forgot_password.ForgotPasswordActivity
import com.settlex.android.presentation.auth.register.RegisterActivity
import com.settlex.android.presentation.common.components.BiometricAuthManager
import com.settlex.android.presentation.common.components.BiometricAuthManager.BiometricAuthCallback
import com.settlex.android.presentation.common.extensions.gone
import com.settlex.android.presentation.common.extensions.show
import com.settlex.android.presentation.common.state.UiState
import com.settlex.android.presentation.common.util.DialogHelper
import com.settlex.android.presentation.common.util.KeyboardHelper
import com.settlex.android.presentation.common.util.SpannableTextFormatter
import com.settlex.android.presentation.dashboard.DashboardActivity
import com.settlex.android.util.string.StringFormatter
import com.settlex.android.util.ui.ProgressDialogManager
import com.settlex.android.util.ui.StatusBar
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class LoginActivity : AppCompatActivity() {

    private lateinit var binding: ActivityLoginBinding
    private val keyboardHelper: KeyboardHelper by lazy { KeyboardHelper(this) }
    private val progressLoader: ProgressDialogManager by lazy { ProgressDialogManager(this) }
    private val viewModel: LoginViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)

        initViews()
        initObservers()
        syncUserStateWithUI()
        initListeners()
    }

    private fun initViews() = with(binding) {
        StatusBar.setColor(this@LoginActivity, R.color.colorBackground)
        setupInputValidation()
        keyboardHelper.attachDoneAction(editText = etPassword)

        tvSwitchAccount.text = SpannableTextFormatter(
            this@LoginActivity,
            "Not you?\nSwitch Account",
            "Switch Account",
            ContextCompat.getColor(this@LoginActivity, R.color.colorPrimary)
        )

        tvSignUp.text = SpannableTextFormatter(
            this@LoginActivity,
            "Don't have an account yet?\nClick here to register",
            "Click here to register",
            ContextCompat.getColor(this@LoginActivity, R.color.colorPrimary)
        )
    }

    private fun initObservers() {
        observeLoginEvent()
    }

    private fun initListeners() = with(binding) {
        tvForgotPassword.setOnClickListener {
            launchActivity(
                ForgotPasswordActivity::class.java
            )
        }
        btnBackBefore.setOnClickListener { onBackPressedDispatcher.onBackPressed() }
        tvSwitchAccount.setOnClickListener { showLoggedOutView() }
        ivFingerprint.setOnClickListener { authenticateWithBiometrics() }
        tvSignUp.setOnClickListener { launchActivity(RegisterActivity::class.java) }
        btnSignIn.setOnClickListener { tryLogin() }
    }

    private fun syncUserStateWithUI() {
        val state = viewModel.userState.value
        when (state) {
            is LoginState.LoggedOut -> showLoggedOutView()
            is LoginState.LoggedInUser -> {
                showLoggedUser(state)

                with(binding) {
                    // Enable/disable fingerprint auth
                    val isFingerPrintEnabled = viewModel.isLoginBiometricsEnabled.value
                    when (isFingerPrintEnabled) {
                        true -> {
                            ivFingerprint.show()
                            authenticateWithBiometrics()
                        }

                        false -> ivFingerprint.gone()
                    }
                }
            }
        }
    }

    private fun showLoggedUser(user: LoginState.LoggedInUser) {
        with(binding) {
            val formattedDisplayName = "Hi, ${user.displayName.uppercase()}"
            val formattedEmail = "(${StringFormatter.maskEmail(user.email)})"

            loadProfilePhoto(user.photoUrl, ivUserProfilePhoto)
            tvUserDisplayName.text = formattedDisplayName
            tvUserEmail.text = formattedEmail
            etEmail.setText(user.email)

            viewAuthenticatedUi.show()
            tvSwitchAccount.show()

            listOf(ivLogo, tilEmail, tvSignUp).forEach { it.gone() }
        }
    }

    private fun showLoggedOutView() = with(binding) {
        viewModel.logout()

        listOf(
            viewAuthenticatedUi,
            ivFingerprint,
            tvSwitchAccount
        ).forEach { it.gone() }

        listOf(
            tilEmail,
            tvSignUp,
            ivLogo
        ).forEach { it.show() }
    }

    // Observers
    private fun observeLoginEvent() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.loginEvent.collect {
                    when (it) {
                        is UiState.Loading -> progressLoader.show()
                        is UiState.Success -> onLoginSuccess()
                        is UiState.Failure -> onLoginFailure(it.exception)
                    }
                }
            }
        }
    }

    private fun onLoginSuccess() {
        startActivity(
            Intent(
                this,
                DashboardActivity::class.java
            )
        )
        finishAffinity()

        progressLoader.hide()
    }

    private fun onLoginFailure(error: AppException) = with(binding) {
        when (error) {
            is AppException.NetworkException -> showNetworkErrorDialog(error)
            else -> {
                tvLoginError.text = error.message
                tvLoginError.show()
            }
        }

        progressLoader.hide()
    }

    private fun showNetworkErrorDialog(error: AppException) {
        DialogHelper.showAlertDialogMessage(
            this
        ) { dialog, binding ->

            with(binding) {
                tvMessage.text = error.message
                "Okay".also { btnPrimary.text = it }

                btnPrimary.setOnClickListener { dialog.dismiss() }
                btnSecondary.gone()
            }
        }
    }

    private fun tryLogin() = with(binding) {
        val email = etEmail.text.toString().trim()
        val password = etPassword.text.toString().trim()

        viewModel.login(email, password)
    }

    private fun authenticateWithBiometrics() {
        if (BiometricAuthManager.isBiometricAvailable(this)) {
            val biometric = BiometricAuthManager(
                this, this, object : BiometricAuthCallback {
                    override fun onAuthenticated() {
                        launchActivity(DashboardActivity::class.java)
                    }

                    override fun onError(message: String?) {}
                    override fun onFailed() {}
                })
            biometric.authenticate("Confirm your identity", "Use Password")
        }
    }

    private fun launchActivity(activityClass: Class<out Activity>) {
        startActivity(Intent(this, activityClass))
    }

    private fun updateSignInButtonState() = with(binding) {
        val email = etEmail.text.toString().trim()
        val password = etPassword.text.toString().trim()
        val isEmailValid = Patterns.EMAIL_ADDRESS.matcher(email).matches()

        btnSignIn.isEnabled = isEmailValid && password.isNotEmpty()
    }

    private fun setupInputValidation() = with(binding) {
        val validationWatcher: TextWatcher = object : TextWatcher {
            override fun beforeTextChanged(
                s: CharSequence?,
                start: Int,
                count: Int,
                after: Int
            ) {
            }

            override fun afterTextChanged(s: Editable?) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                tvLoginError.gone()
                updateSignInButtonState()
            }
        }

        etEmail.addTextChangedListener(validationWatcher)
        etPassword.addTextChangedListener(validationWatcher)
    }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        if (keyboardHelper.handleOutsideTouch(event)) return true
        return super.dispatchTouchEvent(event)
    }
}
